### ISSUE TITLE
Strategy name in conjunction with NamingStrategy

### ISSUE THREAD

**Issue by BreyndotEchse** [zendframework/zend-hydrator#46]
## Problem
Zend-Hydrate does not work as expected when I use strategies and NamingStrategy together. The problem is, that the strategies have to be bound to the _property name_ when you `hydrate` an object with data. But when you `extract` data from an object, the strategies have to be bound to the _array key_. The only exception is ClassMethods-Hydrator which works slightly different. This is a flaw in my opinion as every hydrator should handle strategies and naming strategies identically (see #45).

## Possible solutions
1. Strategies must be bound to the _property name_. AbstractHydrator::hydrateValue und AbstractHydrator::extractValue get called with the _property name_ (NamingStrategyInterface::hydrate) provided.
2. Strategies must be bound to the _array key_. AbstractHydrator::hydrateValue und AbstractHydrator::extractValue get called with the _array key_ (NamingStrategyInterface::extract) provided.
3. Strategies can either be bound to the _property name_ or _array key_. AbstractHydrator::hydrateValue und AbstractHydrator::extractValue get called twice: At first with the _property name_ (NamingStrategyInterface::hydrate) provided. Then again with the _array key_ (NamingStrategyInterface::extract) provided.

I would go with the first approach because objects have the stricter structure and I usually add strategies to cast an array value to a specific type or create an object out of the value. The array structure may change faster than the class.

Feel free to comment and don't hesitate to correct me, if I have missed an important point.

## Examples
Pseudo examples with `Zend\Hydrator\NamingStrategy\UnderScoreNamingStrategy`. This is how hydrators work (except ClassMethods, see above):
### Hydration
```php
$arrayKey = 'foo_bar';
$propertyName = $hydrator->getNamingStrategy()->hydrate($arrayKey); //fooBar
$hydrator->hydrateValue($propertyName, $data[$arrayKey], $data); //Hydrate value with strategy named equal to $propertyName => 'fooBar'
```
### Extraction
```php
$propertyName = 'fooBar';
$arrayKey = $hydrator->getNamingStrategy()->extract($propertyName); //foo_bar
$hydrator->extractValue($arrayKey, $object->$propertyName, $object); //Extract value with strategy named equal to $arrayKey => 'foo_bar'
```
### Flaw
```php
$hydrator->addNamingStrategy(new UnderscoreNamingStrategy());
$hydrator->addStrategy('fooBar', new PseudoStrategy());
$extracted = $hydrator->extract($hydrator->hydrate($object, $data));
/*
$extracted['foo_bar'] and $data['foo_bar'] will not be identical because PseudoStrategy::extract was never called.
You have to bind the same strategy to both, "fooBar" and "foo_bar", to get the expected result.
$hydrator->addStrategy('fooBar', $strategy);
$hydrator->addStrategy('foo_bar', $strategy);
*/
```

**Comment by boesing** [comment]
Having the same issue with one of my objects.
Since we're using `UnderscoreNamingStrategy`, we have to specify even the hydrated and the original "key" having the Strategy implemented.

Created a PR for @BreyndotEchse but actually I'd rather like to drop the naming strategy for hydration and just during extraction, the name might be extracted to ensure that all strategies are set matching the input `$data` array.

e.g. 

```
$data = ['foo_bar_baz' => 'bar'];
$hydrator->addStrategy('foo_bar_baz', new WhateverStrategy());
$object = $hydrator->hydrate($data, new MyObject());
$extracted = $hydrator->extract($object);
```

The User should not check the object or whatever to detect the correct identifier which should be used for strategy detection at all.
The User only must know whats the incoming data array key, and thats usually the case.

But maybe I'm wrong, just a suggestion.

**Comment by boesing** [comment]
Mh. Wondering why this does not get any attention at all. 😕

**Comment by Ocramius** [comment]
Generally lack of time. If you got someone else interested in this issue,
drag them into it. The CR team will get to it when they get to it.

**Comment by boesing** [comment]
Thanks for the insights, Marco. Just wondering how stuff like this or bugs in other repositories aren't even recognized. Dont want to bother.

**Comment by froschdesign** [comment]
@BreyndotEchse 

> 1. Strategies must be bound to the property name. AbstractHydrator::hydrateValue und AbstractHydrator::extractValue get called with the property name (NamingStrategyInterface::hydrate) provided.

Will this work without changing this existing unit test: [`HydratorStrategyTest::testWhenUsingUnderscoreSeparatedKeysHydratorStrategyIsAlwaysConsideredUnderscoreSeparatedToo()`](https://github.com/zendframework/zend-hydrator/blob/master/test/HydratorStrategyTest.php#L103)?

**Comment by boesing** [comment]
Guess this can be closed since PR #45 was merged?
